core/controllers/base.py

```diff
@@ -274,35 +276,58 @@ def dispatch(self):
             self.redirect('/logout?redirect_url=%s' % self.request.uri)
             return
 
-        if self.payload is not None and self.REQUIRE_PAYLOAD_CSRF_CHECK:
-            try:
-                # If user opens a new tab during signup process, the user_id
-                # parameter is set to None and this causes the signup session
-                # to expire. The code here checks if user is on the signup
-                # page and the user_id is None, if that is the case an exception
-                # is raised which is handled by the frontend by showing a
-                # continue to registration modal.
-                if 'signup' in self.request.uri and not self.user_id:
-                    raise self.UnauthorizedUserException(
-                        'Registration session expired.')
-                csrf_token = self.request.get('csrf_token')
-                if not csrf_token:
-                    raise self.UnauthorizedUserException(
-                        'Missing CSRF token. Changes were not saved. '
-                        'Please report this bug.')
-
-                is_csrf_token_valid = CsrfTokenManager.is_csrf_token_valid(
-                    self.user_id, csrf_token)
-
-                if not is_csrf_token_valid:
-                    raise self.UnauthorizedUserException(
-                        'Your session has expired, and unfortunately your '
-                        'changes cannot be saved. Please refresh the page.')
-            except Exception as e:
-                logging.exception('%s: payload %s', e, self.payload)
-
-                self.handle_exception(e, self.app.debug)
-                return
+        try:
+            # If user opens a new tab during signup process, the user_id
+            # parameter is set to None and this causes the signup session
+            # to expire. The code here checks if user is on the signup
+            # page and the user_id is None, if that is the case an exception
+            # is raised which is handled by the frontend by showing a
+            # continue to registration modal.
+            if 'signup' in self.request.uri and not self.user_id:
+                raise self.UnauthorizedUserException(
+                    'Registration session expired.')
+
+            session_expired_error_msg = (
+                'Your session has expired, and unfortunately your '
+                'changes cannot be saved. Please refresh the page.')
+            missing_csrf_token_error_msg = (
+                'Missing CSRF token. Changes were not saved. '
+                'Please report this bug.')
+
+            for arg in self.request.arguments():
```

Comment:
Reason to iterate over all the args in dispatch -- Since our aim is to validate all the args reaching to a handler class. Initially we validate csrf_token in dispatch method but we are calling validate_normalized_args() inside dispatch method so there may be: 
1. repetition of validation of csrf_token in dispatch() as well as validate_normalized_args() method.
2. Or we can exclude the csrf_token validation in validate_normalized_args() since it is already validated in dispatch method (but as guided by Sean we do not have to skip or exclude any of the arg)

After stating these reasons I feel iterating and collecting all the args in dispatch is one of possibility to overcome this problem.